src/components/text_table/state.rs

```diff
@@ -472,24 +472,25 @@ impl<H: TableComponentHeader> TableComponentState<H> {
 
     /// Updates the position if possible, and if there is a valid change, returns the new position.
     pub fn update_position(&mut self, change: i64, num_entries: usize) -> Option<usize> {
-        if change == 0 {
+        if change == 0
+            || (change > 0 && self.current_scroll_position == num_entries - 1)
+            || (change < 0 && self.current_scroll_position == 0)
+        {
             return None;
         }
 
         let csp: Result<i64, _> = self.current_scroll_position.try_into();
         if let Ok(csp) = csp {
-            let proposed: Result<usize, _> = (csp + change).try_into();
+            let proposed: Result<usize, _> = (csp + change).max(0).try_into();
             if let Ok(proposed) = proposed {
-                if proposed < num_entries {
-                    self.current_scroll_position = proposed;
-                    if change < 0 {
-                        self.scroll_direction = ScrollDirection::Up;
-                    } else {
-                        self.scroll_direction = ScrollDirection::Down;
-                    }
-
-                    return Some(self.current_scroll_position);
+                self.current_scroll_position = proposed.min(num_entries - 1);
```

Comment:
Would using the built-in [`clamp`](https://doc.rust-lang.org/std/cmp/trait.Ord.html#method.clamp) method work here?

e.g.:
```rust
(csp + change).clamp(0, num_entries - 1);
```
(note this doesn't handle when `num_entries` is 0)

Would convey your intent better in my opinion.

Reply:
Yeah it would be better, but it would need a try_into() or similar from `num_entries: usize` to `i64`

Reply:
I did a `[usize] as i64` (which can overflow at about 9.2e18 on 64-bit) and a `[i64] as usize` (which can overflow on 32-bit systems). But in practice it should never overflow.